pkg/operator/controller/canary/controller.go

```diff
@@ -0,0 +1,177 @@
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "canary_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the canary controller.
+//
+// The canary controller will watch the Default IngressController, as well as
+// the canary service, daemonset, and route resources.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		Config: config,
+		client: mgr.GetClient(),
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	// Only trigger a reconcile request for the canary controller via events for the default ingress controller.
+	defaultIcPredicate := predicate.NewPredicateFuncs(func(meta metav1.Object, object runtime.Object) bool {
+		return meta.GetName() == manifests.DefaultIngressControllerName
+	})
+
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}, defaultIcPredicate); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Reconcile ensures that the canary controller's resources
+// are in the desired state.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	result := reconcile.Result{}
+	errors := []error{}
+
+	if err := r.ensureCanaryNamespace(); err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
+	}
+
+	haveDs, daemonset, err := r.ensureCanaryDaemonSet()
+	if err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
+	}
+	if !haveDs {
+		errors = append(errors, fmt.Errorf("failed to get canary daemonset"))
+		return result, utilerrors.NewAggregate(errors)
+	}
+
+	trueVar := true
+	daemonsetRef := metav1.OwnerReference{
+		APIVersion: "apps/v1",
+		Kind:       "daemonset",
+		Name:       daemonset.Name,
+		UID:        daemonset.UID,
+		Controller: &trueVar,
+	}
+
+	haveService, service, err := r.ensureCanaryService(daemonsetRef)
+	if err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
+	}
+	if !haveService {
+		errors = append(errors, fmt.Errorf("failed to get canary service"))
+		return result, utilerrors.NewAggregate(errors)
+	}
+
+	if haveRoute, _, err := r.ensureCanaryRoute(service); err != nil {
```

Comment:
Change the returned values of `ensureCanaryRoute()` to the following if the *route is not being used:
```
(bool, error)
```

Reply:
Is this suggestion really necessary? A follow up PR might need to use a field from the canary route resource. Additionally, this pattern is used for the ingress controller logic https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/controller.go#L596.

Reply:
It's not required, but in this case and the case you reference can cause confusion. If a returned value is not being used, then it should not be returned.